solr/solrj/src/java/org/apache/solr/client/solrj/impl/SolrClientNodeStateProvider.java

```diff
@@ -273,30 +273,31 @@ protected void getRemoteInfo(String solrNode, Set<String> requestedTags, SnitchC
 
       try {
         SimpleSolrResponse rsp = snitchContext.invokeWithRetry(solrNode, CommonParams.METRICS_PATH, params);
+        NamedList<?> metrics = (NamedList<?>) rsp.nl.get("metrics");
 
-        Map m = rsp.nl.asMap(4);
         if (requestedTags.contains(FREEDISK.tagName)) {
-          Object n = Utils.getObjectByPath(m, true, "metrics/solr.node/CONTAINER.fs.usableSpace");
+          Object n = Utils.getObjectByPath(metrics, true, "solr.node/CONTAINER.fs.usableSpace");
           if (n != null) ctx.getTags().put(FREEDISK.tagName, FREEDISK.convertVal(n));
         }
         if (requestedTags.contains(TOTALDISK.tagName)) {
-          Object n = Utils.getObjectByPath(m, true, "metrics/solr.node/CONTAINER.fs.totalSpace");
+          Object n = Utils.getObjectByPath(metrics, true, "solr.node/CONTAINER.fs.totalSpace");
           if (n != null) ctx.getTags().put(TOTALDISK.tagName, TOTALDISK.convertVal(n));
         }
         if (requestedTags.contains(CORES)) {
+          NamedList<?> node = (NamedList<?>) metrics.get("solr.node");
           int count = 0;
-          Map cores = (Map) m.get("metrics");
-          for (Object o : cores.keySet()) {
-            if (o.toString().startsWith("solr.core.")) count++;
+          for (String leafCoreMetricName : new String[]{"lazy", "loaded", "unloaded"}) {
+            Number n = (Number) node.get("CONTAINER.cores." + leafCoreMetricName);
+            if (n != null) count += n.intValue();
```

Comment:
I don't know if this issue was isolated to our use case - but I know we sometimes saw negative values for some of the core counts. Do you want to check if (n != null && n > 0)?

Reply:
I've thought of that issue.  I think the proper place to fix that is at the source of the metric, not here (consumer).